mm2src/coins/eth/eth_swap_v2/eth_maker_swap_v2.rs

```diff
@@ -0,0 +1,523 @@
+            },
+            EthCoinType::Nft { .. } => unreachable!(),
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn refund_maker_payment_v2_timelock_impl(
+        &self,
+        args: RefundMakerPaymentTimelockArgs<'_>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let (token_address, gas_limit) = match &self.coin_type {
+            EthCoinType::Eth => (Address::default(), self.gas_limit_v2.maker.eth_maker_refund_timelock),
+            EthCoinType::Erc20 {
+                platform: _,
+                token_addr,
+            } => (*token_addr, self.gas_limit_v2.maker.erc20_maker_refund_timelock),
+            EthCoinType::Nft { .. } => {
+                return Err(TransactionErr::ProtocolNotSupported(ERRL!(
+                    "NFT protocol is not supported for ETH and ERC20 Swaps"
+                )))
+            },
+        };
+
+        let maker_swap_v2_contract = self
+            .swap_v2_contracts
+            .as_ref()
+            .map(|contracts| contracts.maker_swap_v2_contract)
+            .ok_or_else(|| TransactionErr::Plain(ERRL!("Expected swap_v2_contracts to be Some, but found None")))?;
+        let payment_amount = try_tx_s!(wei_from_big_decimal(&args.amount, self.decimals));
+        let (maker_secret_hash, taker_secret_hash) = match args.tx_type_with_secret_hash {
+            SwapTxTypeWithSecretHash::MakerPaymentV2 {
+                maker_secret_hash,
+                taker_secret_hash,
+            } => (maker_secret_hash, taker_secret_hash),
+            _ => {
+                return Err(TransactionErr::Plain(ERRL!(
+                    "Unsupported swap tx type for timelock refund"
+                )))
+            },
+        };
+        let args = {
+            let taker_address = public_to_address(&Public::from_slice(args.taker_pub));
+            MakerRefundArgs {
+                payment_amount,
+                taker_address,
+                taker_secret: [0u8; 32],
+                taker_secret_hash: try_tx_s!(taker_secret_hash.try_into()),
+                maker_secret_hash: try_tx_s!(maker_secret_hash.try_into()),
+                payment_time_lock: args.time_lock,
+                token_address,
+            }
+        };
+        let data = try_tx_s!(self.prepare_refund_maker_payment_timelock_data(args).await);
+
+        self.sign_and_send_transaction(
+            U256::from(ZERO_VALUE),
+            Action::Call(maker_swap_v2_contract),
+            data,
+            U256::from(gas_limit),
+        )
+        .compat()
+        .await
+    }
+
+    pub(crate) async fn refund_maker_payment_v2_secret_impl(
+        &self,
+        args: RefundMakerPaymentSecretArgs<'_, Self>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let (token_address, gas_limit) = match &self.coin_type {
+            EthCoinType::Eth => (Address::default(), self.gas_limit_v2.maker.eth_maker_refund_secret),
```

Comment:
FYI: in other project I noticed that if they need to use a token_address for the ETH coin they use a special const (0xeeee..e) instead of zero. I guess this to prevent errorneous contract calls

Reply:
Yeah I noticed it during 1inch review.
I dont mind to use a default placeholder for native chain coins, as `0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE` is an recognisable placeholder, which is a widely accepted convention in DeFi and DEX aggregators, making it easier to understand across various platforms. Zero address is usually used to init or destroy tokens, so it could cause misunderstanding.

Then we need to update solidity code when we send native coin to swap smart contract
https://github.com/KomodoPlatform/etomic-swap/blob/5e15641cbf41766cd5b37b4d71842c270773f788/contracts/EtomicSwapMakerV2.sol#L51
Because right now we calculate paymentHash with zero address for eth coin.

We can add constant to sodlitiy code of smart contracts
```solidity
address constant ETH_PLACEHOLDER_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
```

I suppose, as constant 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE is embedded in the bytecode, using it will be as gas-efficient as using address(0). But I would like to check it at least in radhat.

Reply:
ETH_PLACEHOLDER_ADDRESS should not affect gas price.
IMO it could affect gas used but as you showed it did not

Reply:
@dimxy during tests got 'Invalid paymentHash' error on the step when Taker spends maker ETH payment
https://sepolia.etherscan.io/tx/0xf04c7c988bbb07ebab1e7c5b0090984b34a20a6928d3fcfdc1bed0207ddb6e8d

```
---- docker_tests::eth_docker_tests::send_and_spend_maker_payment_eth stdout ----
25 12:22:29, eth_docker_tests:894] All pending transactions have been confirmed. Latest nonce: 251
25 12:22:33, eth_docker_tests:2358] Maker sent ETH payment, tx hash: 2a07d07a5fae7c905f5c879fa7f69a25688d18e0bb1e3c9a308b2271fa0a4ae4
· 2024-09-25 19:22:39 +0700 [ETH] Waiting for confirmations… Confirmed.
25 12:22:40, eth_docker_tests:2371] Taker validated maker ETH payment
25 12:22:40, eth_docker_tests:894] All pending transactions have been confirmed. Latest nonce: 833
25 12:22:47, eth_docker_tests:2385] Taker spent maker ETH payment, tx hash: f04c7c988bbb07ebab1e7c5b0090984b34a20a6928d3fcfdc1bed0207ddb6e8d
· 2024-09-25 19:22:51 +0700 [ETH] Waiting for confirmations… Failed.
thread 'docker_tests::eth_docker_tests::send_and_spend_maker_payment_eth' panicked at 'called `Result::unwrap()` on an `Err` value: "eth:5450] Internal: eth:5450] Tx receipt Receipt { transaction_hash: 0xf04c7c988bbb07ebab1e7c5b0090984b34a20a6928d3fcfdc1bed0207ddb6e8d, transaction_index: 10, block_hash: Some(0xa389355e95def9ce2ccdeeb26b35fab0f741b06cf4ee203e1b13d8b71a43e984), block_number: Some(6756986), from: 0x16e281a9f2e7581269a13e516aa79d6a4a1cd980, to: Some(0xf9000589c66df3573645b59c10aa87594edc318f), cumulative_gas_used: 548329, gas_used: Some(31917), contract_address: None, logs: [], status: Some(0), root: None, logs_bloom: 0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, transaction_type: Some(0), effective_gas_price: Some(478049389820) } status of ETH tx 0xf04c7c988bbb07ebab1e7c5b0090984b34a20a6928d3fcfdc1bed0207ddb6e8d is failed"', mm2src/mm2_main/tests/docker_tests/eth_docker_tests.rs:1330:55
```

Its bcz in solidity, addresses are stored in checksummed format, while the address we are using in Rust, derived from [0xee; 20] or from str `"0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"` etc, is all lowercase (`"0xeeee…eeee"`). Even though both represent the same address, solidity treats the checksummed version differently, and this leads to the Invalid paymentHash error when we compare paymentHash inside smart contract methods.

Lets comeback to zero address, its also can be as an address for eth cases, which means eth coin doesnt have token_address.

Reply:
I am not insisting on using ETH_PLACEHOLDER_ADDRESS, just a FYI.
But we have display_eth_address function which just returns checksum address representation. Shouldn't we use it everywhere?

Reply:
> But we have display_eth_address function which just returns checksum address representation. Shouldn't we use it everywhere?

display_eth_address  is for string representation.

As for `Address` type from ethereum primitives, its bytes will be the same for both cases: from [0xee; 20] and from str "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"


```rust
lazy_static! {
    static ref ETH_PLACEHOLDER_ADDRESS: Address = Address::from_str("0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE").unwrap();
    static ref ETH_PLACEHOLDER_ADDRESS1: Address = Address::from_slice(&[0xee; 20]);
}

// [238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238]
```

Reply:
UPD: as we found out that 1inch uses both zero and 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE addresses for erc20 tokens, lets just leave zero address for payment hash calculation